cvmfs/swissknife_graft.cc

```diff
@@ -0,0 +1,249 @@
+                                            const std::string &file_name) {
+  std::string full_input_path = m_input_file + "/" +
+                                (relative_path.size() ? relative_path : ".") + "/" + file_name;
+  std::string full_output_path;
+  if (m_output_file.size()) {
+    full_output_path = m_output_file + "/" +
+                       (relative_path.size() ? relative_path : ".") + "/" + file_name;
+  }
+  Publish(full_input_path, full_output_path, false, false);
+}
+
+
+int swissknife::CommandGraft::Recurse(const std::string &input_file, const std::string &output_file) {
+  m_output_file = output_file;
+  m_input_file = input_file;
+
+  FileSystemTraversal<CommandGraft> traverser(this, input_file, true);
+  traverser.fn_new_file       = &CommandGraft::FileCallback;
+  traverser.fn_new_dir_prefix = &CommandGraft::DirCallback;
+  traverser.Recurse(input_file);
+  return 1;
+}
+
+
+int swissknife::CommandGraft::Main(const swissknife::ArgumentList &args) {
+  const std::string &input_file = *args.find('i')->second;
+  const std::string output_file = (args.find('o') == args.end()) ? "" : *(args.find('o')->second);
+  m_verbose = args.find('v') != args.end();
+
+  struct stat sbuf;
+  bool output_file_is_dir = output_file.size() &&
+                            (0 == stat(output_file.c_str(), &sbuf)) &&
+                            S_ISDIR(sbuf.st_mode);
+  if (output_file_is_dir && (input_file == "-")) {
+    std::cerr << "Output file (" << output_file << "): Is a directory" << std::endl;
+    return 1;
+  }
+
+  if (input_file != "-") {
+    bool input_file_is_dir = (0 == stat(input_file.c_str(), &sbuf)) &&
+                            S_ISDIR(sbuf.st_mode);
+    if (input_file_is_dir) {
+      if (!output_file_is_dir && output_file.size()) {
+        std::cerr << "Input (" << input_file << ") is a directory but output (" << output_file << ") is not" << std::endl;
+        return 1;
+      }
+      if (m_verbose) {
+        std::cerr << "Recursing into directory " << input_file << std::endl;
+      }
+      return Recurse(input_file, output_file);
+    } else {
+      return Publish(input_file, output_file, output_file_is_dir, false);
+    }
+  }
+  return Publish(input_file, output_file, output_file_is_dir, true);
+}
+
+
+int swissknife::CommandGraft::Publish(const std::string &input_file, const std::string &output_file, bool output_file_is_dir, bool input_file_is_stdin) {
+  if (output_file.size() && m_verbose) {
+    std::cout << "Grafting " << input_file << " to " << output_file << std::endl;
+  } else if (!output_file.size()) {
+    std::cout << "Grafting " << input_file << std::endl;
+  }
+  int fd;
+  if (input_file_is_stdin) {
+    fd = 0;
+  } else {
+    fd = open(input_file.c_str(), O_RDONLY, 0755);
+    if (fd < 0) {
+      std::stringstream ss; ss << "Unable to open input file (" << input_file << ")";
+      perror(ss.str().c_str());
+      return 1;
+    }
+  }
+
+  shash::Any hash(shash::kSha1);
```

Comment:
I'd like to hold off on that until we get the uncompressed files in (and do both compression and hash flag at once).